tests/lang/s02_core/s01_procedures/t02_argument_passing.nim

```diff
@@ -0,0 +1,318 @@
+    12
+
+  proc blockReturn(arg: int): int = arg
+
+  ## Block syntax for procedure calls can also be used as an expression,
+  ## although no in the infix expressions (like `doAseert 12 == <block here>`)
+  block:
+    let res = blockReturn():
+      12
+
+    doAssert res == 12
+
+  block:
+    let res = blockReturn:
+      12
+
+    doAssert res == 12
+
+block block_syntax_do:
+  ## If there are multiple arguments that have to be placed on the
+  ## separate lines this can be achieved using `do:` for
+  ## the arguments.
+  proc blockProc(arg1: int, arg2: int): (int, int) = (arg1, arg2)
+
+  block:
+    let res = blockProc():
+      12
+
+    do:
+      23
+
+    doAssert (12, 23) == res
+
+  block:
+    let res = blockProc:
+      12
+
+    do:
+      20
+
+    doAssert (12, 20) == res
+
+  ## Procedures called with block syntax as also be called using regular
+  ## command or parenthised syntax.
+  doAssert (1, 2) == blockProc(1, 2)
+
+  ## It is also possible to supply only certain arguments.
+  block:
+    let res = blockProc(1123):
+      20
+
+    doAssert (1123, 20) == res
+
+
+  block:
+    let res = blockProc 1:
+      20
+
+    doAssert (1, 20) == res
+
+
+block named_argument_call:
+  proc named(arg1, arg2: int): (int, int) = (arg1, arg2)
+
+  doAssert named(1, 2) == (1, 2)
+  doAssert named(arg1 = 1, arg2 = 2) == (1, 2)
+  doAssert named(arg2 = 2, arg1 = 1) == (1, 2)
+
+
+block default_argument:
+  proc withDefault(arg: int = 12): int =
+    return arg
+
+  doAssert withDefault() == 12
+  doAssert withDefault(23) == 23
+
+  proc withDefaultInfer(arg = 12): int =
+    return arg
+
+  doAssert withDefaultInfer() == 12
+  doAssert withDefaultInfer(30) == 30
+
+
+block named_argument_with_default:
+  ## It is possible to call procedure with default argument value
+  ## using named argument passing even regular ordering would've
+  ## prevented this
+
+
+block default_and_regular_argument:
+  proc withDefault(arg1: int, arg2: string = "123"): (int, string) =
+    return (arg1, arg2)
+
+  doAssert withDefault(1) == (1, "123")
+  doAssert withDefault(2, "1") == (2, "1")
+
+block multiple_default_arguments:
+  proc withDefaultSplit(arg1: int = 1, arg2: int = 2): (int, int) =
+    return (arg1, arg2)
+
+  doAssert withDefaultSplit() == (1, 2)
+  doAssert withDefaultSplit(9) == (9, 2)
+  doAssert withDefaultSplit(3, 4) == (3, 4)
+
+
+  proc withDefaultMerged(arg1, arg2: int = 12): (int, int) = (arg1, arg2)
+
+  doAssert withDefaultMerged() == (12, 12)
+  doAssert withDefaultMerged(9) == (9, 12)
+  doAssert withDefaultMerged(3, 4) == (3, 4)
+
+block complex_expression_as_default:
+  block use_other_argument:
+    proc useOther(arg1: int, arg2: int = arg1): (int, int) = (arg1, arg2)
+
+    doAssert useOther(1) == (1, 1)
+    doAssert useOther(1, 2) == (1, 2)
+
+
+  block use_field_of_other_argument:
+    type
+      Object = object
+        field: int
+
+    proc useOther(obj: Object, arg: int = obj.field): (int, int) = (obj.field, arg)
```

Comment:
Todo, to be noted for later, I think the formal semantics of this should be a lowering to a combination of procedures that takes the arguments, or not if default, and then have code in the body to carry out the expressions to access fields and stuff.